components/architecture.md

```diff
@@ -0,0 +1,147 @@
+***Draft***
+
+## Moby Core Component Architecture
+
+Moby Core is a container platform built for modularity.
+Moby Core uses a component architecture which allows moby components to interact
+with each other with well defined API's in a way that enables components
+to be swapped out for different implementations at packaging time.
+
+Moby Core defines several of its own [components](#moby-core-services) along with
+a default stack which implements those services. Packagers can take Moby Core as
+defined, swap out components for different ones, and even supply completely new
+components not specifically defined in the core architecture to extend
+Moby Core's functionality.
+
+### Status
+
+This architecture is still in heavy development and this document should be considered fluid
+for the time being.
+
+While the v1 HTTP API is considered stable, the components and component APIs are not finalized
+and should NOT be considered stable. This document may contain things that do not even exist yet.
+
+***Use at your own risk. Expect frequent breakages.***
+
+### Components
+
+A component is an isolated set of functionality that is likely useful in its
+own right.
+
+A component:
+- MUST provide a GRPC endpoint with the following services on a node-local socket (such as a unix socket or named pipe):
+	- MUST provide [GRPC healthcheck service](https://github.com/grpc/grpc/blob/master/doc/health-checking.md)
+	- MUST provide [Version service](api/services/component/version/v1)
+	- SHOULD provide [Metrics service](api/services/component/metrics/v1)
+- MAY expose a separate API for external consumption
+- SHOULD use GRPC to publish component-level services
+- SHOULD be easily deployed as a container (the primary packaging format for
+Moby Core, see [Packaging](#Packaging)).
```

Comment:
How many processes do you expect will be involved in providing functionality equivalent to Docker?

If components are separate Go binaries running in their own containers, they will be relatively heavyweight, in terms of size and memory footprint. If covering the Docker feature set involves, say, 10 binaries of 30 MB each, that may be a nonstarter for certain embedded applications. I could also see it being an issue for memory-limited VMs, especially in the cloud. If we are inefficient with resources it could push people away from containerization, or towards more bare-bones solutions like standalone containerd.

Reply:
I agree this is something we'll need to watch out for.
However I would prefer to optimize for this later and I don't think it's strictly an architecture problem.

dockerd itself is only 33M, so I expect components to come in quite a bit smaller. I know GRPC brings a hefty price tag in that regard (around 10MB last I saw), though.

Reply:
> GRPC brings a hefty price tag in that regard (around 10MB last I saw)

Wow. We should file that as a bug. I noticed that we are taking on way too many dependencies to carry pre-http2 Go versions. I'll bring that up at the community meeting.